File: 131-palindrome-partitioning/palindrome-partitioning.cpp
class Solution {
public:
    bool isPalindrome(string s){
        string s2 = s;
        reverse(s2.begin(), s2.end());
        if(s == s2){
            return true;
        }
        return false;
    }
    void GetAllParts(string s, vector<vector<string>> &ans, vector<string> &parts){
        if(s.size() == 0){
            ans.push_back(parts);
            return;
        }
        for(int i = 0; i < s.size(); i++){
            string part = s.substr(0, i + 1);
            if(isPalindrome(part)){
                parts.push_back(part);
                GetAllParts(s.substr(i + 1), ans, parts);
                parts.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        vector<vector<string>> ans;
        vector<string> parts;
        GetAllParts(s, ans, parts);
        return ans;
    }
};

File: 131-palindrome-partitioning/palindrome-partitioning.py
class Solution:
    def isPalindrome(self, s):
        return s == s[::-1]

    def getAllParts(self, s, ans, parts):
        if len(s) == 0:
            ans.append(parts[:])
            return
        for i in range(len(s)):
            part = s[:i + 1]
            if self.isPalindrome(part):
                parts.append(part)
                self.getAllParts(s[i + 1:], ans, parts)
                parts.pop()

    def partition(self, s):
        ans = []
        parts = []
        self.getAllParts(s, ans, parts)
        return ans